### ISSUE TITLE
Legend cut of if using pgf backend

### ISSUE THREAD

**Issue by FrederikLauber** [matplotlib/matplotlib#7073]
Hello,

I am trying to save a graphic in pgf. Unfortunately, my legend is being cut of the figure.
If I save the picture as pdf, png or svg, it is displayed correctly.

[light_yield.pdf](https://github.com/matplotlib/matplotlib/files/463956/light_yield.pdf)

![light_yield](https://cloud.githubusercontent.com/assets/17728264/18387246/ce6f4348-769b-11e6-95df-e39c2d71406d.png)

This is my code:

![snapshot2](https://cloud.githubusercontent.com/assets/17728264/18387417/dafc8638-769c-11e6-854d-89334ec0a44e.png)

``` python
fig, ax1 = plt.subplots()
ax2 = ax1.twinx()
norm=norm,linewidth=2.0)

ax1.set_xlabel(r'$\beta$')
ax1.set_ylabel(r'$\frac{dN_\gamma}{dx} / \mathrm{cm}^{-1} \left( \mathrm{theory} \right)$')
ax1.set_yscale('log')
ax1.set_xlim((0, 1))
ax1.set_ylim((10**2, 10**7))
ax1.grid()
ax2.set_ylabel(r'$\overline{N}_{\mathrm{PE}} \left( \mathrm{simulation} \right)$')
ax2.set_yscale("log", nonposy="clip")
ax2.set_ylim((10**2/scale, 10**7/scale))

ax1.spines['left'].set_color('blue')
ax1.spines['right'].set_color('red')
ax1.yaxis.label.set_color('blue')
ax2.yaxis.label.set_color('red')
ax1.tick_params(axis='y', colors='blue')
ax2.tick_params(axis='y', colors='red')

ln1, = ax1.plot(xtheory, directCherenkov, "b",linewidth=2.0, label="direct Cherenkov")
ln2, = ax1.plot(xtheory, indirectCherenkov, "b:",linewidth=2.0, label="indirect Cherenkov")
ln3, = ax1.plot(xtheory, luminescencelight, "b--",linewidth=3.0, label=r"luminescence, $\frac{dN_\gamma}{dE}= 0.2 \gamma / \mathrm{MeV}$")
ln4, _, _ = ax2.errorbar(xsim_before_split, ysim_before_split, yerr=ysimerr_before_split, fmt='ro', label="total, before detector", linewidth=2.0)
ln5, _, _ = ax2.errorbar(xsim_before[scale_point], ysim_before[scale_point], yerr=ysimerr_before[scale_point], fmt='cp', label="reference point between theory and simulations", linewidth=2.0)
ln6, _, _ = ax2.errorbar(xsim_after, ysim_after, yerr=ysimerr_after, fmt='rs', label="total, after detector", linewidth=2.0)

#create combined legend
lns = [ln1,ln2,ln3,ln4,ln6, ln5]
labs = [l.get_label() for l in lns]
labs[-3] = "total, before detector"
labs[-2] = "total, after detector"
labs[-1] = "reference point between left and right axis"
lgd = ax1.legend(lns, labs, bbox_to_anchor=(0., 1.02, 1., 0.102), loc=3, ncol=1, mode="expand", borderaxespad=0., prop={'size':13})

fig.savefig("./plots/light_yield.png", bbox_inches='tight')
fig.savefig("./plots/light_yield.svg", bbox_inches='tight')
fig.savefig("./plots/light_yield.pdf", bbox_inches='tight')
try:
    fig.savefig("./plots/light_yield.pgf", bbox_extra_artists=(lgd,))
except Exception:
    print "Creating pgf failed"
```

I already tried to add the legend manually for the pgf case but it does not help.

[EDIT]
God darn it, I cannot upload the pgf file because uploading is not supported by the tracker.
Basically, the legend is cut directly above the lowest line.

I added a screenshot showing the problem.

**Comment by tacaswell** [comment]
I would bet dollars to doughnuts the issue is `bbox_inches='tight'` not working properly.  Try saving with out that kwargs (which may require adding a call to `fig.tight_layout()` and / or manually playing with the figure size to get right).

attn @pwuertz

**Comment by FrederikLauber** [comment]
Hi,

I am not using bbox_inches='tight' for the pgf file as the pgf backend does not support it in the first place but even if I remove the parameter from the other saves, the problem remains.

I also tried adding a fig.tight_layout() assuming this might lead to a recalculation of the image size before it is stored. This does not change the problem as well.

PS:
One thing I notice during the generation is this message:
/usr/lib/python2.7/dist-packages/matplotlib/scale.py:100: RuntimeWarning: invalid value encountered in less_equal
  a[a <= 0.0] = 1e-300

but it only pops up once and I am generating 3 figures with my plot (each is stored as a svg,pdf,png and pgf)

**Comment by tacaswell** [comment]
Sorry, did not read carefully enough.  `fig.tight_layout` resizes things to fit in figure.  You probably have to either pass `figsize=(widith, height)` to `plt.subplots` or use `fig.set_size_inches(w, h)` to resize the figure.  You may _then_ have to adjust the size of your main `Axes` in figure fraction to make it a bit smaller (or the same thing will happen with the legend running off the top of the figure).

**Comment by pwuertz** [comment]
Unfortunately I'm quite busy at the moment, not sure when I can dive into this. There is a simple check though. Your legend is clipped at the top, so if you look inside the PGF output and find negative y-coordinates for some of your legend items the error is most probably in the layout, not the backend.

**Comment by jklymak** [comment]
The solution for this is to use outside legends for constrained layout.  The other backends were relying on bbox_inches=tight and that's not available on the pgf backend.  https://matplotlib.org/stable/tutorials/intermediate/legend_guide.html#figure-legends
